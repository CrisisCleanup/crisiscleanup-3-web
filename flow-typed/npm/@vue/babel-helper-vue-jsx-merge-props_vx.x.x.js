// flow-typed signature: d7a23c53f4091ef2bf1df8afa5eb43c4
// flow-typed version: <<STUB>>/@vue/babel-helper-vue-jsx-merge-props_v^1.0.0/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-helper-vue-jsx-merge-props'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-helper-vue-jsx-merge-props' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper.js' {
  declare module.exports: $Exports<'@vue/babel-helper-vue-jsx-merge-props/dist/helper'>;
}
