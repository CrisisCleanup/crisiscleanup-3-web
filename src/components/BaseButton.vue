<template>
  <button
    :class="[styles, buttonSelector]"
    :alt="alt || text"
    :disabled="disabled || loading"
    type="submit"
    :size="size"
    :title="buttonTitle"
    :data-cy="buttonSelector"
    @click.prevent="performAction"
  >
    <font-awesome-icon
      v-if="loading || showSpinner"
      size="sm"
      icon="spinner"
      spin
    />
    <font-awesome-icon v-if="icon" class="m-1" :icon="icon" :size="iconSize" />
    <ccu-icon v-if="ccuIcon" class="m-1" :type="ccuIcon" :size="iconSize" />
    <slot>{{ text }}</slot>
    <font-awesome-icon v-if="suffixIcon" class="m-1" :icon="suffixIcon" />
  </button>
</template>

<script>
import { kebabCase } from 'lodash';
import VueTypes from 'vue-types';
import { BUTTON_VARIANTS as VARIANTS, ICON_SIZES, ICONS } from '@/constants';
import { EventsMixin } from '@/mixins';
import { delay } from '@/utils/promise';

export default {
  name: 'BaseButton',
  mixins: [EventsMixin],
  props: {
    action: VueTypes.func,
    disabled: VueTypes.bool.def(false),
    showSpinner: VueTypes.bool.def(false),
    text: VueTypes.string,
    alt: VueTypes.string.def('button'),
    title: VueTypes.string.def('Button'),
    size: VueTypes.oneOf(['small', 'medium', 'large', 'sm', 'md', 'lg']),
    icon: VueTypes.string,
    ccuIcon: VueTypes.oneOf(Object.values(ICONS)),
    iconSize: VueTypes.oneOf(ICON_SIZES).def('sm'),
    suffixIcon: VueTypes.oneOf(Object.values(ICONS)),
    selector: VueTypes.string,
    variant: VueTypes.oneOf(Object.values(VARIANTS)),
  },
  data() {
    return {
      loading: false,
    };
  },
  computed: {
    buttonSelector() {
      return this.selector || `js-${kebabCase(this.buttonTitle)}`;
    },
    buttonTitle() {
      return this.text || this.alt || this.title;
    },
    styles() {
      const styles = {
        'large text-h3 font-h3': ['large', 'lg'].includes(this.size),
        'medium text-sans text-h4 font-h4': ['medium', 'md'].includes(
          this.size,
        ),
        'small text-bodysm font-bodysm': ['small', 'sm'].includes(this.size),
        disabled: this.disabled,
        // **** DEPRECATED ****
        primary: this.type === 'primary',
        danger: this.type === 'danger',
        warning: this.type === 'warning',
        link: this.type === 'link',
        bare: this.type === 'bare',
        // **************
        flex: true,
        'items-center': true,
        'justify-center': true,
        'base-button': true,
      };
      if (this.variant) {
        styles[this.variant] = true;
      }
      styles[this.variant] = true;
      return styles;
    },
  },
  methods: {
    async timeout() {
      await delay(5000);
      return Promise.reject();
    },
    async performAction() {
      this.loading = true;

      try {
        if (this.action) {
          await Promise.race([this.action(), this.timeout()]);
        }
      } catch (e) {
        // TODO: expose method for handling button exceptions
      } finally {
        this.logEvent();
        this.loading = false;
      }
    },
  },
};
</script>

<style scoped>
button {
  @apply font-sans;
  cursor: pointer;
  outline: 0;
  transition: all 300ms ease;
}
button:focus {
  outline: 0;
}

/** ----- DEPRECATED ----- */
.primary {
  @apply bg-primary-light;
}
.link {
  @apply text-primary-dark;
}
/** ----- DEPRECATED ----- */

button.solid {
  @apply bg-primary-light;
}

button.solid:hover {
  @apply bg-crisiscleanup-yellow-100;
}

button.solid.disabled {
  @apply bg-crisiscleanup-dark-200 !important;
  @apply text-gray-100;
}

button.outline {
  background-color: transparent;
  border: 1px solid black;
}

button.outline:hover {
  @apply bg-crisiscleanup-dark-500;
  @apply text-gray-100;
}

button.outline.disabled {
  @apply text-crisiscleanup-grey-900;
}

button.outline.disabled:hover {
  @apply text-crisiscleanup-grey-900;
}

button.outline-dark {
  background-color: transparent;
  border: 1px solid white;
}

button.outline-dark:hover {
  @apply bg-crisiscleanup-dark-500;
  @apply text-gray-100;
}

button.outline-dark.disabled {
  @apply text-crisiscleanup-grey-900;
}

button.outline-dark.disabled:hover {
  @apply text-crisiscleanup-grey-900;
}

button.text {
  background-color: transparent;
}

button.text:hover {
  @apply bg-crisiscleanup-grey-100;
}

button.text.disabled {
  @apply text-crisiscleanup-grey-900;
}

button.text.disabled:hover {
  @apply text-crisiscleanup-grey-900;
}

button.small {
  min-height: 24px;
  padding-left: 15px;
  padding-right: 15px;
}

button.medium {
  min-height: 32px;
  padding-left: 20px;
  padding-right: 20px;
}

button.large {
  min-height: 50px;
  padding-left: 34px;
  padding-right: 34px;
}
</style>
