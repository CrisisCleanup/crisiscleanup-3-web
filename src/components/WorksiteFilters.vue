<template>
  <modal
    v-if="show"
    modal-classes="bg-white max-w-2xl shadow"
    modal-style="min-height: 60%"
  >
    <div class="flex flex-col h-full">
      <div class="p-3">
        {{ $t('worksiteFilters.filters') }}
        <span
          v-if="filtersCount > 0"
          class="rounded-full px-1 bg-yellow-500 text-xs"
          >{{ filtersCount }}</span
        >
      </div>
      <div
        v-if="filtersCount > 0"
        class="
          flex
          items-center
          justify-between
          bg-crisiscleanup-light-grey
          p-1
          px-2
        "
      >
        <div
          class="
            applied-filters
            flex flex-wrap
            justify-start
            bg-crisiscleanup-light-grey
          "
        >
          <template v-for="(filter, key) in filters">
            <template v-for="(label, identifier) in filter.labels">
              <tag
                :key="key + identifier"
                closeable
                class="m-1"
                @closed="
                  () => {
                    filter.removeField(identifier);
                  }
                "
              >
                {{ label }}
              </tag>
            </template>
          </template>
        </div>
        <div>
          <base-button
            type="bare"
            :text="$t('actions.clear_filters')"
            :alt="$t('actions.clear_filters')"
            class="text-yellow-500 text-underline w-32"
            :action="clearAllFilters"
          />
        </div>
      </div>

      <div class="flex flex-grow h-full">
        <div class="w-1/4 border-r">
          <div
            class="p-3 px-4 border-b cursor-pointer"
            :class="{
              'border-l-4 border-l-black': currentSection === 'general',
            }"
            @click="currentSection = 'general'"
          >
            {{ $t('worksiteFilters.general') }}
            <span
              v-if="statusCount + statusGroupCount > 0"
              class="rounded-full px-1 bg-black text-white text-xs"
              >{{ statusCount + statusGroupCount }}</span
            >
          </div>
          <div class="p-3 px-4 border-b cursor-pointer">
            {{ $t('worksiteFilters.personal_info') }}
          </div>
          <div
            class="p-3 px-4 border-b cursor-pointer"
            :class="{
              'border-l-4 border-l-black': currentSection === 'flags',
            }"
            @click="currentSection = 'flags'"
          >
            {{ $t('worksiteFilters.flags') }}
            <span
              v-if="flagsCount > 0"
              class="rounded-full px-1 bg-black text-white text-xs"
              >{{ flagsCount }}</span
            >
          </div>
          <div
            class="p-3 px-4 border-b cursor-pointer"
            :class="{ 'border-l-4 border-l-black': currentSection === 'work' }"
            @click="currentSection = 'work'"
          >
            {{ $t('worksiteFilters.work') }}
            <span
              v-if="fieldsCount + missingWorkTypeCount > 0"
              class="rounded-full px-1 bg-black text-white text-xs"
              >{{ fieldsCount + missingWorkTypeCount }}</span
            >
          </div>
          <div
            class="p-3 px-4 border-b cursor-pointer"
            :class="{ 'border-l-4 border-l-black': currentSection === 'teams' }"
            @click="currentSection = 'teams'"
            v-if="teams.length"
          >
            {{ $t('worksiteFilters.teams') }}
            <span
              v-if="teamsCount > 0"
              class="rounded-full px-1 bg-black text-white text-xs"
              >{{ teamsCount }}</span
            >
          </div>
          <div
            class="p-3 px-4 border-b cursor-pointer"
            :class="{
              'border-l-4 border-l-black': currentSection === 'locations',
            }"
            @click="currentSection = 'locations'"
          >
            {{ $t('worksiteFilters.locations') }}
            <span
              v-if="locationsCount > 0"
              class="rounded-full px-1 bg-black text-white text-xs"
              >{{ locationsCount }}</span
            >
          </div>
        </div>
        <div
          class="w-3/4 ml-4 mt-2 flex-grow"
          style="height: 450px; overflow: auto"
        >
          <div v-if="currentSection === 'general'" class="flex flex-col">
            <div class="claim-status mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.location') }}
              </div>
              <base-checkbox
                v-model="
                  filters.locations.data['organization_primary_location']
                "
                class="block my-1"
              >
                {{ $t('worksiteFilters.in_primary_response_area') }}
              </base-checkbox>
              <base-checkbox
                v-model="
                  filters.locations.data['organization_secondary_location']
                "
                class="block my-1"
              >
                {{ $t('worksiteFilters.in_secondary_response_area') }}
              </base-checkbox>
            </div>
            <div class="claim-status mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.team') }}
              </div>
              <base-checkbox
                v-model="filters.my_team.data.my_team"
                class="block my-1"
              >
                {{ $t('worksiteFilters.assigned_to_my_team') }}
              </base-checkbox>
            </div>
            <div class="claim-status mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.claim_reported_by') }}
              </div>
              <base-checkbox
                v-model="filters.statusGroups.data['unclaimed']"
                class="block my-1"
                >{{ $t('worksiteFilters.unclaimed') }}
              </base-checkbox>
              <base-checkbox
                v-model="filters.statusGroups.data['claimed_by_org']"
                class="block my-1"
                >{{ $t('worksiteFilters.claimed_by_my_org') }}
              </base-checkbox>
              <base-checkbox
                v-model="filters.statusGroups.data['reported_by_org']"
                class="block my-1"
                >{{ $t('worksiteFilters.reported_by_my_org') }}
              </base-checkbox>
            </div>
            <div class="status-group mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.over_all_status') }}
              </div>
              <base-checkbox
                class="block my-1"
                :value="filters.statusGroups.data['open']"
                @input="
                  (value) => {
                    setOpenClosed(value, 'open');
                  }
                "
                >{{ $t('worksiteFilters.open') }}
              </base-checkbox>
              <base-checkbox
                class="block my-1"
                :value="filters.statusGroups.data['closed']"
                @input="
                  (value) => {
                    setOpenClosed(value, 'closed');
                  }
                "
                >{{ $t('worksiteFilters.closed') }}
              </base-checkbox>
            </div>

            <div class="statuses mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.detailed_status') }}
              </div>
              <div
                v-for="status in allStatuses"
                :key="`${status.id}`"
                :value="status.status"
              >
                <base-checkbox
                  class="block my-1"
                  :value="filters.statuses.data[status.status]"
                  @input="
                    (value) => {
                      filters.statuses.data[status.status] = value;
                      filters.statuses.data = {
                        ...filters.statuses.data,
                      };
                    }
                  "
                  >{{ status.status | getStatusName }}
                </base-checkbox>
              </div>
            </div>
          </div>
          <div v-if="currentSection === 'flags'" class="flex flex-col">
            <div class="status-group mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.flags') }}
              </div>
              <base-checkbox
                v-for="flag in flagTypes"
                :key="flag"
                class="block my-1"
                :value="filters.flags.data[flag]"
                @input="
                  (value) => {
                    filters.flags.data[flag] = value;
                    filters.flags.data = { ...filters.flags.data };
                  }
                "
                >{{ $t(flag) }}
              </base-checkbox>
            </div>
          </div>
          <template v-if="currentSection === 'work'">
            <div
              v-for="f in incidentTypes"
              :key="f.key"
              :header="f.name_t"
              class="p-2 px-4 mb-2 bg-crisiscleanup-light-grey"
            >
              <div class="flex items-center justify-between">
                <base-checkbox v-model="filters.fields.data[f.key]">{{
                  f.name_t
                }}</base-checkbox>
                <font-awesome-icon
                  v-if="filters.fields.data[f.key]"
                  class="cursor-pointer"
                  size="md"
                  :icon="expanded[f.key] ? 'caret-up' : 'caret-down'"
                  @click="expandSection(f.key)"
                />
              </div>
              <div v-if="expanded[f.key]">
                <template v-for="field in getFieldsForType(f.key)">
                  <div :key="field.field_key" class="border-b py-3">
                    <template v-if="field.html_type === 'select'">
                      <div class="font-bold">
                        {{ field.label_t }}
                      </div>
                      <div>
                        <div>
                          <div
                            v-for="option in field.values.filter((option) =>
                              Boolean(option.value),
                            )"
                            :key="option.value"
                            :span="8"
                          >
                            <base-checkbox :value="option.value">
                              {{ option.name_t }}
                            </base-checkbox>
                          </div>
                        </div>
                      </div>
                    </template>
                    <template v-if="field.html_type === 'multiselect'">
                      <div class="font-bold">
                        {{ field.label_t }}
                      </div>
                      <div>
                        <div>
                          <div
                            v-for="option in field.values.filter((option) =>
                              Boolean(option.value),
                            )"
                            :key="option.value"
                            :span="8"
                          >
                            <base-checkbox :value="option.value">
                              {{ option.name_t }}
                            </base-checkbox>
                          </div>
                        </div>
                      </div>
                    </template>
                    <template v-if="field.html_type === 'checkbox'">
                      <div class="flex">
                        <span class="font-bold w-1/2">
                          {{ field.label_t }}
                        </span>
                        <div class="flex justify-around w-1/2">
                          <base-checkbox>{{
                            $t('worksiteFilters.yes')
                          }}</base-checkbox>
                          <base-checkbox>{{
                            $t('worksiteFilters.no')
                          }}</base-checkbox>
                          <base-checkbox>{{
                            $t('worksiteFilters.maybe')
                          }}</base-checkbox>
                        </div>
                      </div>
                    </template>
                  </div>
                </template>
              </div>
            </div>
            <div class="status-group mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.missing_information') }}
              </div>
              <base-checkbox
                v-model="filters.missingWorkType.data['missing_work_type']"
                class="block my-1"
              >
                {{ $t('worksiteFilters.no_work_type') }}
              </base-checkbox>
            </div>
          </template>
          <div v-if="currentSection === 'teams'" class="flex flex-col">
            <div class="status-group mb-2">
              <div class="my-1 text-base">
                {{ $t('worksiteFilters.teams') }}
              </div>
              <base-checkbox
                v-for="team in teams"
                :key="`${team.id}`"
                class="block my-1"
                :value="filters.teams.data[team.id]"
                @input="
                  (value) => {
                    filters.teams.data[team.id] = value;
                    filters.teams.data = { ...filters.teams.data };
                  }
                "
                >{{ team.name }}
              </base-checkbox>
            </div>
          </div>
          <div v-if="currentSection === 'locations'" class="flex flex-col">
            <div class="mb-2">
              <div class="claim-status mb-2" v-if="filters.locations">
                <div class="my-1 text-base">
                  {{ $t('worksiteFilters.response_areas') }}
                </div>
                <base-checkbox
                  v-model="
                    filters.locations.data['organization_primary_location']
                  "
                  class="block my-1"
                >
                  {{ $t('worksiteFilters.in_primary_response_area') }}
                </base-checkbox>
                <base-checkbox
                  v-model="
                    filters.locations.data['organization_secondary_location']
                  "
                  class="block my-1"
                >
                  {{ $t('worksiteFilters.in_secondary_response_area') }}
                </base-checkbox>
              </div>
              <div class="mb-2">
                <div class="my-1 text-base">
                  {{ $t('worksiteFilters.my_locations') }}
                </div>
                <div v-for="location in locations" :key="`${location.id}`">
                  <base-checkbox
                    class="block my-1"
                    :value="filters.locations.data[location.id]"
                    @input="
                      (value) => {
                        filters.locations.data[location.id] = value;
                        filters.locations.data = {
                          ...filters.locations.data,
                        };
                      }
                    "
                    >{{ location.name }}
                  </base-checkbox>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      slot="footer"
      class="flex items-center justify-center p-2 bg-white border-t"
    >
      <base-button
        :text="$t('actions.cancel')"
        :alt="$t('actions.cancel')"
        size="medium"
        class="m-1 border-2 border-black px-6 py-2"
        :action="
          () => {
            $emit('closedFilters');
          }
        "
      />
      <base-button
        :text="$t('actions.apply_filters')"
        :alt="$t('actions.apply_filters')"
        ccu-event="user_ui-turn-on_filter"
        size="medium"
        class="m-1 p-3 px-6"
        variant="solid"
        :action="updateFilters"
      />
    </div>
  </modal>
</template>

<script>
import { mapState } from 'vuex';
import Team from '@/models/Team';
import WorksiteFieldsFilter from '../utils/data_filters/WorksiteFieldsFilter';
import WorksiteFlagsFilter from '../utils/data_filters/WorksiteFlagsFilter';
import WorksiteStatusGroupFilter from '../utils/data_filters/WorksiteStatusGroupFilter';
import WorksiteStatusFilter from '../utils/data_filters/WorksiteStatusFilter';
import WorksiteLocationsFilter from '../utils/data_filters/WorksiteLocationsFilter';
import WorksiteMissingWorkTypeFilter from '../utils/data_filters/WorksiteMissingWorkTypeFilter';
import WorksiteMyTeamFilter from '../utils/data_filters/WorksiteMyTeamFilter';
import WorksiteTeamsFilter from '../utils/data_filters/WorksiteTeamsFilter';
import { UserMixin } from '../mixins';

export default {
  name: 'WorksiteFilters',
  mixins: [UserMixin],
  props: {
    incident: {
      type: Object,
      default: () => {
        return {};
      },
    },
    currentFilters: {
      type: Object,
      default: () => {
        return {};
      },
    },
    locations: {
      type: Array,
      default: () => {
        return [];
      },
    },
    show: {
      type: Boolean,
    },
  },
  data() {
    return {
      filters: {
        fields: {},
        statuses: {},
        teams: {},
        my_team: {},
        statusGroups: {},
        flags: {},
        sub_fields: {},
        locations: {},
        missingWorkType: {},
      },
      currentSection: 'general',
      expanded: {},
      flagTypes: [
        'flag.worksite_high_priority',
        'flag.worksite_upset_client',
        'flag.worksite_mark_for_deletion',
        'flag.worksite_abuse',
        'flag.duplicate',
        'flag.worksite_wrong_location',
        'flag.worksite_wrong_incident',
      ],
    };
  },
  computed: {
    incidentTypes() {
      if (this.incident && this.incident.form_fields) {
        const fieldsWithTypes = this.incident.form_fields.filter((field) => {
          return Boolean(field.if_selected_then_work_type);
        });

        const types = new Set(
          fieldsWithTypes.map((t) => t.if_selected_then_work_type),
        );

        return Array.from(types).map((workType) => {
          return this.workTypes.find((type) => type.key === workType);
        });
      }
      return [];
    },
    fieldsCount() {
      return this.filters.fields.count;
    },
    statusCount() {
      return this.filters.statuses.count;
    },
    statusGroupCount() {
      return this.filters.statusGroups.count;
    },
    flagsCount() {
      return this.filters.flags.count;
    },
    locationsCount() {
      return this.filters.locations.count;
    },
    missingWorkTypeCount() {
      return this.filters.missingWorkType.count;
    },
    teamsCount() {
      return this.filters.teams.count;
    },
    myTeamCount() {
      return this.filters.my_team.count;
    },
    filtersCount() {
      return (
        this.fieldsCount +
        this.statusCount +
        this.statusGroupCount +
        this.locationsCount +
        this.flagsCount +
        this.missingWorkTypeCount +
        this.teamsCount +
        this.myTeamCount
      );
    },
    teams() {
      return Team.all();
    },
    ...mapState('enums', ['statuses', 'workTypes']),
    allStatuses() {
      return this.statuses.map((status, index) => {
        return {
          ...status,
          selectionKey: index + 1,
        };
      });
    },
  },
  watch: {
    currentFilters() {
      this.filters = {
        fields: new WorksiteFieldsFilter(
          'fields',
          (this.currentFilters.fields && this.currentFilters.fields.data) || {},
        ),
        statusGroups: new WorksiteStatusGroupFilter(
          'statusGroups',
          (this.currentFilters.statusGroups &&
            this.currentFilters.statusGroups.data) ||
            {},
        ),
        flags: new WorksiteFlagsFilter(
          'flags',
          (this.currentFilters.flags && this.currentFilters.flags.data) || {},
        ),
        locations: new WorksiteLocationsFilter(
          'locations',
          (this.currentFilters.locations &&
            this.currentFilters.locations.data) ||
            {},
        ),
        statuses: new WorksiteStatusFilter(
          'statuses',
          (this.currentFilters.statuses && this.currentFilters.statuses.data) ||
            {},
        ),
        my_team: new WorksiteMyTeamFilter(
          'my_team',
          (this.currentFilters.my_team && this.currentFilters.my_team.data) ||
            {},
        ),
        teams: new WorksiteTeamsFilter(
          'teams',
          (this.currentFilters.teams && this.currentFilters.teams.data) || {},
        ),
        missingWorkType: new WorksiteMissingWorkTypeFilter(
          'missingWorkType',
          (this.currentFilters.missingWorkType &&
            this.currentFilters.missingWorkType.data) ||
            {},
        ),
      };
    },
  },
  methods: {
    updateFilters() {
      this.$emit('updatedFilters', {
        ...this.filters,
      });

      // eslint-disable-next-line no-restricted-syntax
      for (const [key, value] of Object.entries(this.filters.fields.data)) {
        if (!value) {
          this.expanded[key] = false;
        }
      }
    },
    /* eslint-enable no-restricted-syntax */
    setOpenClosed(value, status) {
      this.filters.statusGroups.data.open = false;
      this.filters.statusGroups.data.closed = false;
      if (value) {
        this.filters.statusGroups.data[status] = value;
      }
      this.filters.statusGroups.data = {
        ...this.filters.statusGroups.data,
      };
    },
    expandSection(key) {
      this.expanded[key] = !this.expanded[key];
      this.expanded = { ...this.expanded };
    },
    setSubFields(workType, field, values) {
      this.sub_fields[workType][field] = values;
    },
    getFieldsForType(workType) {
      if (this.incident && this.incident.form_fields) {
        return this.incident.form_fields.filter((field) => {
          const parent = this.incident.form_fields.find((element) => {
            return element.field_key === field.field_parent_key;
          });

          // eslint-disable-next-line camelcase
          let { if_selected_then_work_type } = field;
          if (parent) {
            if_selected_then_work_type = parent.if_selected_then_work_type;
          }
          return if_selected_then_work_type === workType;
        });
      }
      return [];
    },
    clearAllFilters() {
      this.filters = {
        fields: new WorksiteFieldsFilter('fields', {}),
        statusGroups: new WorksiteStatusGroupFilter('statusGroups', {}),
        flags: new WorksiteFlagsFilter('flags', {}),
        statuses: new WorksiteStatusFilter('statuses', {}),
        locations: new WorksiteLocationsFilter('locations', {}),
        teams: new WorksiteTeamsFilter('teams', {}),
        my_team: new WorksiteMyTeamFilter('my_team', {}),
        missingWorkType: new WorksiteMissingWorkTypeFilter(
          'missingWorkType',
          {},
        ),
      };
    },
  },
};
</script>

<style scoped></style>
