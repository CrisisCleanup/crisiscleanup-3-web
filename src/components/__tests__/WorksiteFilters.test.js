/**
 *
 * Tests for WorksiteFilters
 *
 * Components
 */

import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import BaseButton from '@/components/BaseButton';
import { snakeToTitleCase } from '@/filters';
import WorksiteFilters from '../WorksiteFilters';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.filter('getWorkTypeName', () => 'workType.ash');
localVue.filter('snakeToTitleCase', snakeToTitleCase);

jest.mock('@/models/Worksite');

const MockFilters = {
  fields: {
    muck_out: true,
  },
  statuses: {
    open: true,
    closed: false,
  },
  sub_fields: {},
};

const mocks = {
  $t: key => key,
};

const mountWithOptions = () =>
  shallowMount(WorksiteFilters, {
    stubs: {
      'base-button': BaseButton,
      'base-checkbox': true,
      modal: true,
      tag: true,
    },
    propsData: {
      currentFilters: MockFilters,
    },
    localVue,
    mocks,
  });

describe('WorksiteFilters', () => {
  it('should not log any errors', () => {
    const spy = jest.spyOn(global.console, 'error');
    mountWithOptions();
    expect(spy).not.toHaveBeenCalled();
  });

  it('should set status open/closed', () => {
    const wrapper = mountWithOptions();
    wrapper.vm.setOpenClosed(true, 'open');
    expect(wrapper.vm.filters).toMatchSnapshot();
    wrapper.vm.setOpenClosed();
    expect(wrapper.vm.filters).toMatchSnapshot();
  });

  it('should remove field', () => {
    const wrapper = mountWithOptions();
    expect(wrapper.vm.filters).toMatchSnapshot(MockFilters);
    wrapper.vm.removeField('muck_out');
    expect(wrapper.vm.filters).toMatchSnapshot({
      ...MockFilters,
      fields: {},
    });
  });

  it('should remove status', () => {
    const wrapper = mountWithOptions();
    wrapper.vm.removeStatus('open');
    expect(wrapper.vm.filters).toMatchSnapshot({
      ...MockFilters,
      statuses: {
        closed: false,
      },
    });
  });

  it('should update filters', () => {
    const wrapper = mountWithOptions();
    wrapper.vm.updateFilters();
    const expectFilters = MockFilters;
    delete expectFilters.sub_fields;
    expect(wrapper.emitted().updatedFilters[0][0]).toMatchSnapshot(
      expectFilters,
    );
  });

  it('should render correctly and match snapshot', () => {
    const wrapper = mountWithOptions();
    expect(wrapper.element).toMatchSnapshot();
  });
});
