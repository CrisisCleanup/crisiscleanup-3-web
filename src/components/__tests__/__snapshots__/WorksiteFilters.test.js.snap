// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorksiteFilters should remove field 1`] = `
Object {
  "fields": Object {
    "muck_out": true,
  },
  "statusGroups": Object {
    "closed": false,
    "open": true,
  },
  "statuses": Object {},
  "sub_fields": Object {},
}
`;

exports[`WorksiteFilters should remove field 2`] = `
Object {
  "fields": Object {
    "muck_out": undefined,
  },
  "statusGroups": Object {
    "closed": false,
    "open": true,
  },
  "statuses": Object {},
  "sub_fields": Object {},
}
`;

exports[`WorksiteFilters should remove status 1`] = `
Object {
  "fields": Object {
    "muck_out": true,
  },
  "statusGroups": Object {
    "closed": false,
    "open": undefined,
  },
  "statuses": Object {
    "open": undefined,
  },
  "sub_fields": Object {},
}
`;

exports[`WorksiteFilters should render correctly and match snapshot 1`] = `
<modal-stub
  modal-classes="bg-white max-w-2xl shadow"
  modal-style="min-height: 60%"
>
  <div
    class="flex flex-col h-full"
  >
    <div
      class="p-3"
    >
      
      Filters
      
      <span
        class="rounded-full px-1 bg-yellow-500 text-xs"
      >
        2
      </span>
    </div>
     
    <div
      class="flex items-center justify-between bg-gray-100 p-1 px-2"
    >
      <div
        class="applied-filters flex flex-wrap justify-start bg-gray-100"
      >
        <tag-stub
          class="m-1"
          closeable=""
        >
          
            worksiteFilters.work_type: workType.ash
          
        </tag-stub>
         
        <tag-stub
          class="m-1"
          closeable=""
        >
          
            worksiteFilters.status: Open
          
        </tag-stub>
        <!---->
         
      </div>
       
      <div>
        <button
          alt="actions.clear_filters"
          class="text-yellow-500 text-underline w-32 bare flex items-center justify-center"
          size=""
          title="actions.clear_filters"
          type="submit"
        >
          <!---->
           
          <!---->
           actions.clear_filters 
          <!---->
        </button>
      </div>
    </div>
     
    <div
      class="flex flex-grow h-full"
    >
      <div
        class="w-1/4 border-r"
      >
        <div
          class="p-3 px-4 border-b cursor-pointer border-l-8 border-l-black"
        >
          
          General
          
          <span
            class="rounded-full px-1 bg-black text-white text-xs"
          >
            1
          </span>
        </div>
         
        <div
          class="p-3 px-4 border-b cursor-pointer"
        >
          
          Personal Info
        
        </div>
         
        <div
          class="p-3 px-4 border-b cursor-pointer"
        >
          
          Work
          
          <span
            class="rounded-full px-1 bg-black text-white text-xs"
          >
            1
          </span>
        </div>
      </div>
       
      <div
        class="w-3/4 ml-4 mt-2 flex-grow"
        style="height: 450px; overflow: auto;"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="claim-status mb-2"
          >
            <div
              class="my-1 text-base"
            >
              Claimed or Reported by
            </div>
             
            <base-checkbox-stub
              class="block my-1"
            >
              worksiteFilters.unclaimed
            
            </base-checkbox-stub>
             
            <base-checkbox-stub
              class="block my-1"
            >
              worksiteFilters.claimed_by_my_org
            
            </base-checkbox-stub>
             
            <base-checkbox-stub
              class="block my-1"
            >
              worksiteFilters.reported_by_my_org
            
            </base-checkbox-stub>
          </div>
           
          <div
            class="status-group mb-2"
          >
            <div
              class="my-1 text-base"
            >
              Over-all Status
            </div>
             
            <base-checkbox-stub
              class="block my-1"
              value="true"
            >
              worksiteFilters.open
            
            </base-checkbox-stub>
             
            <base-checkbox-stub
              class="block my-1"
            >
              worksiteFilters.closed
            
            </base-checkbox-stub>
          </div>
           
          <div
            class="statuses mb-2"
          >
            <div
              class="my-1 text-base"
            >
              Detailed Status
            </div>
             
            <div
              value="open_unassigned"
            >
              <base-checkbox-stub
                class="block my-1"
              >
                Open, unassigned
              
              </base-checkbox-stub>
            </div>
            <div
              value="open_assigned"
            >
              <base-checkbox-stub
                class="block my-1"
              >
                Open, unassigned
              
              </base-checkbox-stub>
            </div>
          </div>
        </div>
         
        <!---->
      </div>
    </div>
  </div>
   
  <div
    class="flex items-center justify-center p-2 bg-white border-t"
  >
    <button
      alt="actions.cancel"
      class="m-1 border-2 border-black px-6 py-2 text-base flex items-center justify-center"
      size="medium"
      title="actions.cancel"
      type="submit"
    >
      <!---->
       
      <!---->
       actions.cancel 
      <!---->
    </button>
     
    <button
      alt="actions.apply_filters"
      class="m-1 p-3 px-6 text-base primary flex items-center justify-center"
      size="medium"
      title="actions.apply_filters"
      type="submit"
    >
      <!---->
       
      <!---->
       actions.apply_filters 
      <!---->
    </button>
  </div>
</modal-stub>
`;

exports[`WorksiteFilters should set status open/closed 1`] = `
Object {
  "fields": Object {
    "muck_out": true,
  },
  "statusGroups": Object {
    "closed": false,
    "open": true,
  },
  "statuses": Object {},
  "sub_fields": Object {},
}
`;

exports[`WorksiteFilters should set status open/closed 2`] = `
Object {
  "fields": Object {
    "muck_out": true,
  },
  "statusGroups": Object {
    "closed": false,
    "open": false,
  },
  "statuses": Object {},
  "sub_fields": Object {},
}
`;

exports[`WorksiteFilters should update filters 1`] = `
Object {
  "fields": Object {
    "muck_out": true,
  },
  "statusGroups": Object {
    "closed": false,
    "open": true,
  },
  "statuses": Object {},
}
`;
